File: app/src/main/java/ru/p3tr0vich/fuel/FuelingAdapter.java
package ru.p3tr0vich.fuel;

import android.databinding.DataBindingUtil;
import android.support.annotation.IntDef;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

import ru.p3tr0vich.fuel.databinding.FuelingListItemBinding;
import ru.p3tr0vich.fuel.models.FuelingRecord;

class FuelingAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private static final int TYPE_HEADER = 0;
    private static final int TYPE_ITEM = 1;
    public static final int TYPE_FOOTER = 2;

    private static final long HEADER_ID = Long.MAX_VALUE;
    private static final long FOOTER_ID = Long.MIN_VALUE;

    public static final int HEADER_POSITION = 0;

    private final List<FuelingRecord> mFuelingRecords;

    private boolean mShowYear;

    @HeaderFooter
    private int mShowHeader;
    @HeaderFooter
    private int mShowFooter;

    @IntDef({HF_HIDE, HF_SHOW})
    public @interface HeaderFooter {
    }

    private static final int HF_HIDE = 0;
    private static final int HF_SHOW = 1;

    private final View.OnClickListener mOnClickListener;

    FuelingAdapter(View.OnClickListener onClickListener,
                   boolean showHeader, boolean showFooter) {
        super();
        setHasStableIds(true);

        setShowHeader(showHeader);
        setShowFooter(showFooter);

        mFuelingRecords = new ArrayList<>();
        if (isShowHeader()) mFuelingRecords.add(null);
        if (isShowFooter()) mFuelingRecords.add(null);

        mOnClickListener = onClickListener;
    }

    public boolean isShowHeader() {
        return mShowHeader == HF_SHOW;
    }

    private void setShowHeader(boolean showHeader) {
        mShowHeader = showHeader ? HF_SHOW : HF_HIDE;
    }

    private boolean isShowFooter() {
        return mShowFooter == HF_SHOW;
    }

    private void setShowFooter(boolean showFooter) {
        mShowFooter = showFooter ? HF_SHOW : HF_HIDE;
    }

    public void swapRecords(@Nullable List<FuelingRecord> records) {
//        UtilsLog.d("FuelingAdapter", "swapCursor",
//                data == null ? "data == null" : "data count == " + data.getCount());

        mFuelingRecords.clear();

        if (isShowHeader()) mFuelingRecords.add(null);

        if (records != null)
            mFuelingRecords.addAll(records);

        if (isShowFooter()) mFuelingRecords.add(null);

        notifyDataSetChanged();
    }

    public int findPositionById(long id) {
        for (int i = mShowHeader, count = mFuelingRecords.size() - mShowFooter; i < count; i++)
            if (mFuelingRecords.get(i).getId() == id) return i;

        return -1;
    }

    public void setShowYear(boolean showYear) {
        mShowYear = showYear;
    }

    @Override
    public int getItemCount() {
        return mFuelingRecords.size();
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        switch (viewType) {
            case TYPE_ITEM:
                return new FuelingItemViewHolder(
                        FuelingListItemBinding.inflate(
                                LayoutInflater.from(parent.getContext()), parent, false).getRoot());
            case TYPE_HEADER:
                return new HeaderViewHolder(
                        LayoutInflater.from(parent.getContext()).
                                inflate(R.layout.fueling_recycler_view_header, parent, false));
            case TYPE_FOOTER:
                return new HeaderViewHolder(
                        LayoutInflater.from(parent.getContext()).
                                inflate(R.layout.fueling_recycler_view_footer, parent, false));
            default:
                throw new RuntimeException("onCreateViewHolder: wrong viewType == " + viewType);
        }
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        if (holder instanceof FuelingItemViewHolder) {
            final FuelingRecord fuelingRecord = mFuelingRecords.get(position);

            final FuelingListItemBinding binding = ((FuelingItemViewHolder) holder).getBinding();

            binding.setFuelingRecord(fuelingRecord);

            binding.btnMenu.setTag(fuelingRecord.getId());
            binding.btnMenu.setOnClickListener(mOnClickListener);

            binding.setShowYear(mShowYear);
        }
    }

    @Override
    public int getItemViewType(int position) {
        if (isShowHeader() && position == HEADER_POSITION) return TYPE_HEADER;
        if (isShowFooter() && position == mFuelingRecords.size() - 1) return TYPE_FOOTER;
        return TYPE_ITEM;
    }

    @Override
    public long getItemId(int position) {
        if (isShowHeader() && position == HEADER_POSITION) return HEADER_ID;
        if (isShowFooter() && position == mFuelingRecords.size() - 1) return FOOTER_ID;

        return mFuelingRecords.get(position).getId();
    }

    public class FuelingItemViewHolder extends RecyclerView.ViewHolder {

        private final FuelingListItemBinding mBinding;

        public FuelingItemViewHolder(View itemView) {
            super(itemView);
            mBinding = DataBindingUtil.bind(itemView);
        }

        public FuelingListItemBinding getBinding() {
            return mBinding;
        }
    }

    public class HeaderViewHolder extends RecyclerView.ViewHolder {

        public HeaderViewHolder(View itemView) {
            super(itemView);
        }
    }
}

File: app/src/main/kotlin/ru/p3tr0vich/fuel/adapters/FuelingAdapter.kt
package ru.p3tr0vich.fuel.adapters

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.IntDef
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.RecyclerView
import ru.p3tr0vich.fuel.R
import ru.p3tr0vich.fuel.databinding.FuelingListItemBinding
import ru.p3tr0vich.fuel.models.FuelingRecord
import java.util.*

internal class FuelingAdapter(private val onClickListener: View.OnClickListener,
                              showHeader: Boolean, showFooter: Boolean) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private val fuelingRecords: MutableList<FuelingRecord> = ArrayList()

    var showYear: Boolean = false

    @HeaderFooter
    private var mShowHeader = 0
    @HeaderFooter
    private var mShowFooter = 0

    var isShowHeader: Boolean
        get() = mShowHeader == HF_SHOW
        private set(showHeader) {
            mShowHeader = if (showHeader) HF_SHOW else HF_HIDE
        }

    private var isShowFooter: Boolean
        get() = mShowFooter == HF_SHOW
        set(showFooter) {
            mShowFooter = if (showFooter) HF_SHOW else HF_HIDE
        }

    @IntDef(HF_HIDE, HF_SHOW)
    annotation class HeaderFooter

    init {
        setHasStableIds(true)

        isShowHeader = showHeader
        isShowFooter = showFooter

        if (isShowHeader) {
            fuelingRecords.add(FuelingRecord())
        }
        if (isShowFooter) {
            fuelingRecords.add(FuelingRecord())
        }
    }

    fun swapRecords(records: List<FuelingRecord>?) {
        fuelingRecords.clear()

        if (isShowHeader) {
            fuelingRecords.add(FuelingRecord())
        }

        if (records != null) {
            fuelingRecords.addAll(records)
        }

        if (isShowFooter) {
            fuelingRecords.add(FuelingRecord())
        }

        notifyDataSetChanged()
    }

    fun findPositionById(id: Long): Int {
        var i = mShowHeader
        val count = fuelingRecords.size - mShowFooter
        while (i < count) {
            if (fuelingRecords[i].id == id) return i
            i++
        }

        return -1
    }

    override fun getItemCount(): Int {
        return fuelingRecords.size
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return when (viewType) {
            TYPE_ITEM -> FuelingItemViewHolder(
                    FuelingListItemBinding.inflate(
                            LayoutInflater.from(parent.context), parent, false).root)
            TYPE_HEADER -> HeaderViewHolder(
                    LayoutInflater.from(parent.context).inflate(R.layout.partial_fueling_recycler_view_header, parent, false))
            TYPE_FOOTER -> HeaderViewHolder(
                    LayoutInflater.from(parent.context).inflate(R.layout.partial_fueling_recycler_view_footer, parent, false))
            else -> throw RuntimeException("onCreateViewHolder: wrong viewType == $viewType")
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (holder is FuelingItemViewHolder) {
            val fuelingRecord = fuelingRecords[position]

            holder.binding?.fuelingRecord = fuelingRecord

            holder.binding?.btnMenu?.tag = fuelingRecord.id
            holder.binding?.btnMenu?.setOnClickListener(onClickListener)

            holder.binding?.showYear = showYear
        }
    }

    override fun getItemViewType(position: Int): Int {
        if (isShowHeader && position == HEADER_POSITION) return TYPE_HEADER
        return if (isShowFooter && position == fuelingRecords.size - 1) TYPE_FOOTER else TYPE_ITEM
    }

    override fun getItemId(position: Int): Long {
        if (isShowHeader && position == HEADER_POSITION) return HEADER_ID
        return if (isShowFooter && position == fuelingRecords.size - 1) FOOTER_ID else fuelingRecords[position].id

    }

    inner class FuelingItemViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val binding: FuelingListItemBinding? = DataBindingUtil.bind(itemView)
    }

    inner class HeaderViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView)

    companion object {
        private const val TYPE_HEADER = 0
        private const val TYPE_ITEM = 1
        const val TYPE_FOOTER = 2

        private const val HEADER_ID = java.lang.Long.MAX_VALUE
        private const val FOOTER_ID = java.lang.Long.MIN_VALUE

        const val HEADER_POSITION = 0

        private const val HF_HIDE = 0
        private const val HF_SHOW = 1
    }
}